### ISSUE TITLE
GridFieldPaginator() not switching pages

### ISSUE THREAD

**Issue by zanderwar** [silverstripe/silverstripe-framework#6545]
Within `ModelAdmin`:

```php
public function getEditForm($id = null, $fields = null)
{
    /** @var Form $form */
    $form = parent::getEditForm($id, $fields);

    /** @var GridField $gridField */
    $gridField = $form->Fields()->fieldByName($this->sanitiseClassName($this->modelClass));

    $config = GridFieldConfig_RecordEditor::create();
    $config->addComponent(new GridFieldPaginator());
    $gridField->setConfig($config);

    return $form;
}
```

This adds the Pagination controls to the bottom however attempting to navigate through the pages will always refresh the first page and never actually change the page

The POST invoked by the GridField if it's helpful:

```
action_gridFieldAlterAction?StateID=gf_e7311087:Next
filter[-Zanderwar-Scraplands-Models-Player][PlayerName]:
```

**Comment by robbieaverill** [comment]
I can replicate this by adding the paginator component to a GridField. Some notes:

* I had 79 records, and the default page size (without pagination) was 30
* When GridFieldPaginator was added, the pagination controls showed 3 pages (correct)
* It would switch to page 2, but not page 3 (would jump back to page 1 again)
* When I specified 30 as the constructor argument (same as default) it wouldn't leave the first page

**Comment by dhensby** [comment]
Which version of SS?

**Comment by zanderwar** [comment]
SS4 Alpha 4

**Comment by dhensby** [comment]
could it be a regression from my PR: https://github.com/silverstripe/silverstripe-framework/pull/6263?

**Comment by zanderwar** [comment]
Is this: 
```
$lastShownRecord = $state->currentPage * $itemsPerPage;` 
```

meant to be the first record of the current page? otherwise shouldn't it be 
```
$lastShownRecord = ($state->currentPage * $itemsPerPage) - 1;
```

For example if you're on page 3 of 5, showing 10 items per page, wouldn't the last record shown be 29 on page 2? 

However wouldn't explain how that PR would break it since it does the same thing previously

**Comment by anselmdk** [comment]
I can confirm this. I just pulled in the latest version of SS4 framework, and my model admins are not able to paginate (they were't able to do so before either, which is why I updated framework).
In order to enable pagination I added the following lines to my model admin (taken from [this pull request](https://github.com/silverleague/silverstripe-logviewer/pull/12)):

```php
public function getEditForm($id = null, $fields = null)
{
    $form = parent::getEditForm($id, $fields);
    $gridField = $form->Fields()->fieldByName($this->getGridFieldName());
    $config = $gridField->getConfig();
    $config->addComponent(new GridFieldPaginator);
    return $form;
}
public function getGridFieldName()
{
    return $this->sanitiseClassName($this->modelClass);
}
```

Now, when I press the pagination it loads the second page, and on top I'm getting the CMS spinner which won't go away. If I remove it by inspecting it in the browser and press the next page, I'm being sent back to the first page.

**Comment by robbieaverill** [comment]
@dhensy this doesn't appear to be caused by #6263 - I've reverted those changes and the bug persists.

**Comment by robbieaverill** [comment]
Ok - I've done some debugging, and there's actually no problem with the paginator at all.

The Paginator is added by default (`GridFieldConfig_Base::__construct`), but it's not visible at the bottom of the page because it's hanging out off canvas. I tested by adding `margin-bottom: 60px` to the `.field__paginator__controls` element, and the paginator controls are working fine (Chrome/OSX) This CSS doesn't work in IE10 though - it's still off the bottom of the canvas by about half of the element height. I didn't test in any other browsers.

The problem of jumping between pages is caused by having _two paginators in the GridFieldConfig_. Oddly enough, adding the paginator (because you can't see one to start with, so assume there isn't one) makes it show up in the perfect place, it just doesn't work because the list gets paginated twice and confuses the states.

**Comment by anselmdk** [comment]
I can confirm @robbieaverill's findings. I just did the same, and brought the paginator back with some css hacking, I can now paginate. I'm still having the overlay not disappearing though.
See screenshot.
<img width="1061" alt="screenshot 2017-01-28 21 02 40" src="https://cloud.githubusercontent.com/assets/1316533/22399508/f047390a-e59d-11e6-8046-a9649ddc7167.png">

**Comment by zanderwar** [comment]
```
Requirements::customCSS("table.table.grid-field__table {
        margin-bottom: 55px;
}");
```

Fixes it perfectly for me, thanks @robbieaverill

**Comment by robbieaverill** [comment]
@zanderwar cross browser?

**Comment by anselmdk** [comment]
I can confirm that by reverting to `4.0.0-alpha4` and implementing that css hack, pagination now works - without the spinner issue. That must be something having been introduced post `4.0.0-alpha4`.

**Comment by robbieaverill** [comment]
The spinner issue is logged at #5685

**Comment by tractorcow** [comment]
See https://github.com/silverstripe/silverstripe-framework/issues/6585#issuecomment-282595526; I'm having trouble replicating this issue on dev-master. I see the spinner appear as normal, and disappear after loading completes.

Regarding @robbieaverill 's issue:

> The Paginator is added by default (GridFieldConfig_Base::__construct - or GridFieldConfig_RecordEditor for model admins), but it's not visible at the bottom of the page because it's hanging out off canvas.

I've fixed this with https://github.com/silverstripe/silverstripe-framework/pull/6653. Possibly that's all that was necessary?

**Comment by tractorcow** [comment]
Immediately after I posted this, I managed to run into the bug. Oh dear...

**Comment by robbieaverill** [comment]
@tractorcow yeah, there wasn't much to the pagination bug. I hadn't PR'd because I haven't had a chance to work out the best fix for it and didn't want to PR a CSS hack.

**Comment by tractorcow** [comment]
It seems that it works MOST of the time; Trying to figure out what I did that made the bug appear.

**Comment by tractorcow** [comment]
According to @robbieaverill the issue was probably just the addition of the second paginator breaking navigation. Hope that the template CSS class fixes that then. ;)

**Comment by robbieaverill** [comment]
Yep - I'll summarise:

## Initial problem

Not seeing pagination controls in a custom ModelAdmin, I (and others) added a GridFieldPaginator component. Doing so shows pagination controls, but there is already a paginator component, so you end up with two of them.

This causes a bunch of weird stuff to happen, but it is solely caused by having two paginator components applied.

## Actual problem

The default paginator component for model admin GridFields is not visible in the viewport.

Changing the position of it with CSS and testing it, it works fine. It's just initially invisible.

---

@tractorcow's fix at #6653 works to fix this issue.
